lib/Backends/OpenCL/kernels.cl

```diff
@@ -1497,5 +1497,53 @@ __kernel void scatterassignW(__global void *mem, cl_uint32_t data,
   scatterassignK(&mem[data], &mem[indices], &mem[slices], sliceSize);
 }
 
+__kernel void sparselengthsweightedsumK(__global float *dest,
+                                        __global float *data,
+                                        __global float *weights,
+                                        __global cl_uint64_t *indices,
+                                        __global cl_int32_t *lengths,
+                                        cl_uint32_t dataSliceSize) {
+  // Get the global ID. This corresponds to the segment that this kernel will
+  // compute, and therefore also the index into the output buffer at which this
+  // kernel will write its output.
+  cl_int32_t idx = get_global_id(0);
+
+  // Add together the segment lengths of the segments being processed by kernels
+  // with global ID 0 .. idx - 1. This is the index into the indices buffer
+  // at which *this* kernel should start reading indices with which to index
+  // into the data buffer and weights that the data should be multiplied by.
+  cl_int32_t curIdx = 0;
+  for (cl_int32_t i = 0; i < idx; ++i) {
```

Comment:
This loop does the same work multiple times across kernels with different IDs. Is it possible to do this work once on the host side and pass in another buffer?